File: 78.subsets.js
/*
 * @lc app=leetcode id=78 lang=javascript
 *
 * [78] Subsets
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var subsets = function (nums) {
  const res = [[]];
  for (let i = 0; i < nums.length; ++i) {
    const len = res.length;
    for (let j = 0; j < len; ++j) {
      res.push(res[j].concat([nums[i]]));
    }
  }
  return res;
};
// @lc code=end

File: 78.subsets.py
#
# @lc app=leetcode id=78 lang=python3
#
# [78] Subsets
#

# Time:  O()
# Space: O()

from typing import List


# @lc code=start
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        nums.sort()

        # iterative
        res = [[]]
        for i in range(len(nums)):
            size = len(res)
            for j in range(size):
                res.append(list(res[j]))
                res[-1].append(nums[i])
        return res

        # recursive
        def helper(cur: List[int], idx: int):
            res.append(cur.copy())
            for i in range(idx, len(nums)):
                cur.append(nums[i])
                helper(cur, i + 1)
                cur.pop()

        res = []
        helper([], 0)
        return res


# @lc code=end
